File: ballast-client/npm/src/signalr/services/signalr-client-game-service.ts
import { IAddPlayerOptions, ICreateGameOptions, IGameDto, IGameService, IRemovePlayerOptions, IVesselDto, IVesselMoveRequest } from "ballast-core";
import { inject, injectable } from "inversify";
import { TYPES as BallastClient } from "../../dependency-injection/types";
import { ISignalRClientOptions } from "../signalr-client-options";
import { SignalRClientServiceBase } from "../signalr-client-service-base";

@injectable()
export class SignalRClientGameService extends SignalRClientServiceBase implements IGameService {

    public static readonly hubName: string = "gamehub";
    public get hubName() {
        return SignalRClientGameService.hubName;
    }

    public constructor(
        @inject(BallastClient.SignalR.ISignalRClientOptions) serviceOptions: ISignalRClientOptions
    ) {
        super(serviceOptions);
    }
    
    public async getAllGamesAsync(): Promise<IGameDto[]> {
        return await this.createInvocationAsync<IGameDto[]>("GetAllGamesAsync");
    }    

    public async getGameAsync(gameId: string): Promise<IGameDto> {
        return await this.createInvocationAsync<IGameDto>("GetGameAsync", gameId);
    }

    public async getTestGameIdAsync(): Promise<string> {
        return await this.createInvocationAsync<string>("GetTestGameIdAsync");
    }

    public async createGameAsync(options: ICreateGameOptions): Promise<IGameDto> {
        return await this.createInvocationAsync<IGameDto>("CreateGameAsync", options);
    }

    public async startGameAsync(gameId: string): Promise<IGameDto> {
        return await this.createInvocationAsync<IGameDto>("StartGameAsync", gameId);
    }

    public async endGameAsync(gameId: string): Promise<IGameDto> {
        return await this.createInvocationAsync<IGameDto>("EndGameAsync", gameId);
    }

    public async deleteGameAsync(gameId: string): Promise<void> {
        await this.createInvocationAsync("DeleteGameAsync", gameId);
    }

    public async addPlayerToGameAsync(options: IAddPlayerOptions): Promise<IGameDto> {
        return await this.createInvocationAsync<IGameDto>("AddPlayerToGameAsync", options);
    }

    public async removePlayerFromGameAsync(options: IRemovePlayerOptions): Promise<IGameDto> {
        return await this.createInvocationAsync<IGameDto>("RemovePlayerFromGameAsync", options);
    }

    public async addPlayerToVesselAsync(options: IAddPlayerOptions): Promise<IVesselDto> {
        return await this.createInvocationAsync<IVesselDto>("AddPlayerToVesselAsync", options);
    }

    public async removePlayerFromVesselAsync(options: IRemovePlayerOptions): Promise<IVesselDto> {
        return await this.createInvocationAsync<IVesselDto>("RemovePlayerFromVesselAsync", options);
    }

    public async addPlayerToVesselRoleAsync(options: IAddPlayerOptions): Promise<IVesselDto> {
        return await this.createInvocationAsync<IVesselDto>("AddPlayerToVesselRoleAsync", options);
    }

    public async removePlayerFromVesselRoleAsync(options: IRemovePlayerOptions): Promise<IVesselDto> {
        return await this.createInvocationAsync<IVesselDto>("RemovePlayerFromVesselRoleAsync", options);
    }

    public async moveVesselAsync(request: IVesselMoveRequest): Promise<IVesselDto> {
        return await this.createInvocationAsync<IVesselDto>("MoveVesselAsync", request);
    }

}

File: ballast-client/dotnet/src/signalr/services/SignalRClientGameService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Ballast.Core.Application.Models;
using Ballast.Core.Application.Services;

namespace Ballast.Client.SignalR.Services
{
    public class SignalRClientGameService : SignalRClientServiceBase, IGameService
    {

        public SignalRClientGameService(ISignalRClientOptions options) : base(options) { }

        protected override string HubName => "gamehub";

        public Task<GameDto> AddPlayerToGameAsync(AddPlayerOptions options) =>
            CreateInvocationAsync<GameDto>("AddPlayerToGameAsync", options);

        public Task<VesselDto> AddPlayerToVesselAsync(AddPlayerOptions options) =>
            CreateInvocationAsync<VesselDto>("AddPlayerToVesselAsync", options);

        public Task<VesselDto> AddPlayerToVesselRoleAsync(AddPlayerOptions options) =>
            CreateInvocationAsync<VesselDto>("AddPlayerToVesselRoleAsync", options);

        public Task<GameDto> CreateGameAsync(CreateGameOptions options) =>
            CreateInvocationAsync<GameDto>("CreateGameAsync", options);

        public Task DeleteGameAsync(Guid gameId) =>
            CreateInvocationAsync("DeleteGameAsync", gameId);

        public Task<GameDto> EndGameAsync(Guid gameId) =>
            CreateInvocationAsync<GameDto>("EndGameAsync", gameId);

        public Task<IEnumerable<GameDto>> GetAllGamesAsync() =>
            CreateInvocationAsync<IEnumerable<GameDto>>("GetAllGamesAsync");

        public Task<GameDto> GetGameAsync(Guid gameId) =>
            CreateInvocationAsync<GameDto>("GetGameAsync", gameId);

        public Task<Guid> GetTestGameIdAsync() =>
            CreateInvocationAsync<Guid>("GetTestGameIdAsync");

        public Task<VesselDto> MoveVesselAsync(VesselMoveRequest request) =>
            CreateInvocationAsync<VesselDto>("MoveVesselAsync", request);

        public Task<GameDto> RemovePlayerFromGameAsync(RemovePlayerOptions options) =>
            CreateInvocationAsync<GameDto>("RemovePlayerFromGameAsync", options);

        public Task<VesselDto> RemovePlayerFromVesselAsync(RemovePlayerOptions options) =>
            CreateInvocationAsync<VesselDto>("RemovePlayerFromVesselAsync", options);

        public Task<VesselDto> RemovePlayerFromVesselRoleAsync(RemovePlayerOptions options) =>
            CreateInvocationAsync<VesselDto>("RemovePlayerFromVesselRoleAsync", options);

        public Task<GameDto> StartGameAsync(Guid gameId) =>
            CreateInvocationAsync<GameDto>("StartGameAsync", gameId);
        
    }
}


        
